src/webpub_manifest_parser/core/registry.py

```diff
@@ -0,0 +1,158 @@
+
+
+class RegistryItem(object):
+    """Single metadata registry item (collection role, media type, etc.)."""
+
+    def __init__(self, key):
+        """Initialize a new instance of RegistryItem class.
+
+        :param key: Unique identifier of this registry item
+        :type key: str
+        """
+        self._key = key
+
+    @property
+    def key(self):
+        """Return a unique identifier of this registry item.
+
+        :return: Unique identifier of this registry item
+        :rtype: str
+        """
+        return self._key
+
+
+class Registry(collections.MutableMapping):
+    """Collection of registry items with a particular type (collection roles, media types, etc.)."""
+
+    def __init__(self, items=None):
+        """Initialize a new instance of Registry class.
+
+        :param items: (Optional) collection of registry items. Note that all items have to be RegistryItem descendants
+        :type items: List[RegistryItems]
+        """
+        self._items = {}
+
+        if items:
+            self._add_items(items)
+
+    def __setitem__(self, key, value):
+        """Add a new item to the registry.
+
+        :param key: Unique identifier of the item
+        :type key: str
+
+        :param value: Registry item
+        :type value: RegistryItem
+        """
+        if not isinstance(value, RegistryItem):
+            raise ValueError("Registry item must have RegistryItem type")
+
+        self._items[key] = value
+
+    def __delitem__(self, key):
+        """Remove the item from the registry.
+
+        :param key: Unique identifier of the item
+        :type key: str
+        """
+        del self._items[key]
+
+    def __getitem__(self, key):
+        """Return an item from the registry by its key or raises a KeyError if it doesn't exist.
+
+        :param key: Unique identifier of the item
+        :type key: str
+
+        :return: Registry item
+        :rtype: RegistryItem
+        """
+        return self._items[key]
+
+    def __iter__(self):
+        """Return an iterator of all the registry items.
+
+        :return: Iterator of the registry items
+        :rtype: Iterator[RegistryItem]
+        """
+        return iter(self._items.values())
+
+    def __len__(self):
+        """Return a number of items in the registry.
+
+        :return: Number of items in the registry
+        :rtype: int
+        """
+        return len(self._items)
+
+    def _add_items(self, items):
+        """Add new items to the registry. Note that all the items must be RegistryItem descendants.
+
+        :param items: New registry items
+        :type items: List[RegistryItem]
+        """
+        for item in items:
+            if not isinstance(item, RegistryItem):
+                raise ValueError("Registry item must have RegistryItem type")
+
+            self._items[item.key] = item
+
+
+class MediaType(RegistryItem):
+    """Registry item representing a specific media type."""
+
+
+class LinkRelation(RegistryItem):
```

Comment:
[I recently added in Swift a `LinkRelation` struct](https://github.com/readium/r2-shared-swift/blob/5d74f1487d15da06c2a130fb9fde667f64a4aeba/r2-shared-swift/Publication/LinkRelation.swift#L10) to hold the various known relations and have `rels` be type safe. Since it needs to be opened to extensions, this is not an enum.

Reply:
I see, I tried to split all different types or link relations into separate repositories. For example, in the case of RWPM there is a separate [RWPMLinkRelationsRegistry](https://github.com/vbessonov/python-webpub-manifest-parser/blob/feature/new-parser/src/webpub_manifest_parser/rwpm/registry.py#L35):
```
class RWPMLinkRelationsRegistry(Registry):
    """Registry containing link relations mentioned in the RWPM spec."""

    ALTERNATE = LinkRelation(key="alternate")
    CONTENTS = LinkRelation(key="contents")
    COVER = LinkRelation(key="cover")
    MANIFEST = LinkRelation(key="manifest")
    SEARCH = LinkRelation(key="search")
    SELF = LinkRelation(key="self")

    CORE_LINK_RELATIONS = [ALTERNATE, CONTENTS, COVER, MANIFEST, SEARCH, SELF]

    def __init__(self):
        """Initialize a new instance of RWPMLinkRelationsRegistry class."""
        super(RWPMLinkRelationsRegistry, self).__init__(self.CORE_LINK_RELATIONS)
```

Reply:
For OPDS 2.0 there is [OPDS2LinkRelationsRegistry](https://github.com/vbessonov/python-webpub-manifest-parser/blob/feature/new-parser/src/webpub_manifest_parser/opds2/registry.py#L27):
```
class OPDS2LinkRelationsRegistry(RWPMLinkRelationsRegistry):
    """Registry containing OPDS 2.0 link relations."""

    ACQUISITION = LinkRelation(key="http://opds-spec.org/acquisition")
    OPEN_ACCESS = LinkRelation(key="http://opds-spec.org/acquisition/open-access")
    BORROW = LinkRelation(key="http://opds-spec.org/acquisition/borrow")
    BUY = LinkRelation(key="http://opds-spec.org/acquisition/buy")
    SAMPLE = LinkRelation(key="http://opds-spec.org/acquisition/sample")
    PREVIEW = LinkRelation(key="preview")
    SUBSCRIBE = LinkRelation(key="http://opds-spec.org/acquisition/subscribe")

    CORE_LINK_RELATIONS = [
        ACQUISITION,
        OPEN_ACCESS,
        BORROW,
        BUY,
        SAMPLE,
        PREVIEW,
        SUBSCRIBE,
    ]

    def __init__(self):
        """Initialize a new instance of OPDS2LinkRelationsRegistry class."""
        super(OPDS2LinkRelationsRegistry, self).__init__()

        self._add_items(self.CORE_LINK_RELATIONS)
```

Reply:
It looks like it fits a similar need indeed.